src/twisted/internet/test/test_asyncioreactor.py

```diff
@@ -48,8 +56,107 @@ def completed(future):
+        self._test_reactor(reactor)
+        set_event_loop_policy(None)
+
+
+    @skipIf(_defaultEventLoopIsSelector,
+            "default event loop: {}\nis of type SelectorEventLoop "
+            "on Python {}.{} ({})".format(
+                type(_defaultEventLoop), sys.version_info.major,
+                sys.version_info.minor, platform.getType()))
+    def test_defaultNotASelectorEventLoopFromGlobalPolicy(self):
+        """
+        On Windows Python 3.5 to 3.7, L{get_event_loop()} returns a
+        L{WindowsSelectorEventLoop} by default.
+        On Windows Python 3.8+, L{get_event_loop()} returns a
+        L{WindowsProactorEventLoop} by default.
+        L{AsyncioSelectorReactor} should raise a
+        L{ValueError} if the default event loop is not a
+        L{WindowsSelectorEventLoop}.
+        """
+        self.assertRaises(ValueError, AsyncioSelectorReactor)
+
+
+    @skipIf(not hasWindowsProactorEventLoopPolicy,
+            "WindowsProactorEventLoop not available")
+    def test_WindowsProactorEventLoop(self):
+        """
+        L{AsyncioSelectorReactor} will raise a L{ValueError}
+        if instantiated with a L{asyncio.WindowsProactorEventLoop}
+        """
+        event_loop = WindowsProactorEventLoopPolicy().new_event_loop()
+        self.assertRaises(ValueError, AsyncioSelectorReactor, event_loop)
+
+
+    @skipIf(not hasWindowsSelectorEventLoopPolicy,
+            "WindowsSelectorEventLoop only on Windows")
+    def test_WindowsSelectorEventLoop(self):
+        """
+        L{WindowsSelectorEventLoop} works with L{AsyncioSelectorReactor}
+        """
+        event_loop = WindowsSelectorEventLoopPolicy().new_event_loop()
+        reactor = AsyncioSelectorReactor(event_loop)
+        set_event_loop(event_loop)
+        self._test_reactor(reactor)
+        set_event_loop_policy(None)
+
+
+    @skipIf(not hasWindowsProactorEventLoopPolicy,
+            "WindowsProactorEventLoopPolicy only on Windows")
+    def test_WindowsProactorEventLoopPolicy(self):
+        """
+        L{AsyncioSelectorReactor} will raise a L{ValueError}
+        if L{asyncio.WindowsProactorEventLoopPolicy} is default.
+        """
+        set_event_loop_policy(WindowsProactorEventLoopPolicy())
+        with self.assertRaises(ValueError):
+            AsyncioSelectorReactor()
+        set_event_loop_policy(None)
+
+
+    @skipIf(not hasWindowsSelectorEventLoopPolicy,
+            "WindowsSelectorEventLoopPolicy only on Windows")
+    def test_WindowsSelectorEventLoopPolicy(self):
+        """
+        L{AsyncioSelectorReactor} will work if
+        if L{asyncio.WindowsSelectorEventLoopPolicy} is default.
+        """
+        set_event_loop_policy(WindowsSelectorEventLoopPolicy())
+        reactor = AsyncioSelectorReactor()
+        self._test_reactor(reactor)
+        set_event_loop_policy(None)
+
+
     def test_seconds(self):
         """L{seconds} should return a plausible epoch time."""
+        if hasWindowsSelectorEventLoopPolicy:
```

Comment:
I think that all the tests using _test_reactor should e moved into a separate test case.
And the `AsyncioSelectorReactorTests` test case can have a setup method in which it will set WindowsSelectorEventLoopPolicy and reset on teardown.

Reply:
Done

Reply:
I agree, but I don't think it's necessary to hold up the PR for this.

Reply:
It should at least have been a context manager or `try`/`finally` since presently it will just not reset the policy for any failing test...  plus I'm already going to go copy/paste this stuff over into #1263 just to catch up.